File: core/src/main/java/org/elkd/core/consensus/LeaderContext.java
package org.elkd.core.consensus;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import org.elkd.core.server.cluster.Node;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class LeaderContext {
  @VisibleForTesting static final long DEFAULT_MATCH_INDEX = -1;

  private final Map<Node, Long> mNextIndex = new HashMap<>();
  private final Map<Node, Long> mMatchIndex = new HashMap<>();

  public LeaderContext(final Set<Node> nodes, final long lastLogIndex) {
    Preconditions.checkNotNull(nodes, "nodes");

    for (final Node node : nodes) {
      mNextIndex.put(node, lastLogIndex + 1);
      mMatchIndex.put(node, DEFAULT_MATCH_INDEX);
    }
  }

  long getNextIndex(final Node node) {
    Preconditions.checkNotNull(node, "node");
    Preconditions.checkState(mNextIndex.containsKey(node), "node not found in context");

    return mNextIndex.get(node);
  }

  long getMatchIndex(final Node node) {
    Preconditions.checkNotNull(node, "node");
    Preconditions.checkState(mMatchIndex.containsKey(node), "node not found in context");

    return mMatchIndex.get(node);
  }

  void updateNextIndex(final Node node, final long index) {
    Preconditions.checkNotNull(node, "node");
    Preconditions.checkState(mNextIndex.containsKey(node), "node not found in context");

    mNextIndex.put(node, index);
  }

  void updateMatchIndex(final Node node, final long index) {
    Preconditions.checkNotNull(node, "node");
    Preconditions.checkState(mMatchIndex.containsKey(node));

    mMatchIndex.put(node, index);
  }

  @Override
  public String toString() {
    return "LeaderContext{" +
        "mNextIndex=" + mNextIndex +
        ", mMatchIndex=" + mMatchIndex +
        '}';
  }
}

File: core/src/main/java/org/elkd/core/consensus/LeaderContext.kt
package org.elkd.core.consensus

import com.google.common.base.Preconditions
import org.elkd.core.server.cluster.Node
import org.elkd.shared.annotations.Mockable
import java.util.*

@Mockable
class LeaderContext(nodes: Set<Node>, lastLogIndex: Long) {
  private val nextIndex = HashMap<Node, Long>()
  private val matchIndex = HashMap<Node, Long>()

  init {
    for (node in nodes) {
      nextIndex[node] = lastLogIndex + 1
      matchIndex[node] = DEFAULT_MATCH_INDEX
    }
  }

  fun getNextIndex(node: Node): Long {
    Preconditions.checkNotNull(node, "node")
    Preconditions.checkState(node in nextIndex, "node not found in context")

    return nextIndex[node]!! /* node is guaranteed to be in this set */
  }

  fun getMatchIndex(node: Node): Long {
    Preconditions.checkNotNull(node, "node")
    Preconditions.checkState(node in matchIndex, "node not found in context")

    return matchIndex[node]!! /* node is guaranteed to be in this set */
  }

  fun updateNextIndex(node: Node, index: Long) {
    Preconditions.checkNotNull(node, "node")
    Preconditions.checkState(node in nextIndex, "node not found in context")

    nextIndex[node] = index
  }

  fun updateMatchIndex(node: Node, index: Long) {
    Preconditions.checkNotNull(node, "node")
    Preconditions.checkState(node in matchIndex)

    matchIndex[node] = index
  }

  override fun toString(): String {
    return "LeaderContext(nextIndex=$nextIndex, matchIndex=$matchIndex)"
  }

  companion object {
    const val DEFAULT_MATCH_INDEX = 0L
  }
}
